ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsFetch.scala

```diff
@@ -289,85 +291,112 @@ private class ContractsFetch(
+
+              case (AbsoluteBookmark(_), _) | (LedgerBegin, true) =>
+                val stepsAndOffset = builder add transactionsFollowingBoundary(txnK)
+                stepsAndOffset.in <~ Source.single(startOffset)
+                (
+                  (stepsAndOffset: FanOutShape2[_, ContractStreamStep.LAV1, _]).out0,
+                  stepsAndOffset.out1,
+                )
+            }
 
-          val transactInsertsDeletes = Flow
-            .fromFunction(
-              jsonifyInsertDeleteStep(
-                (_: InsertDeleteStep[Any, lav1.event.CreatedEvent]),
-                templateId,
+            val transactInsertsDeletes = Flow
+              .fromFunction(
+                jsonifyInsertDeleteStep(
+                  (_: InsertDeleteStep[Any, lav1.event.CreatedEvent]),
+                  templateId,
+                )
               )
-            )
-            .via(if (sjd.q.queries.allowDamlTransactionBatching) conflation else Flow.apply)
-            .map(insertAndDelete)
+              .via(if (sjd.q.queries.allowDamlTransactionBatching) conflation else Flow.apply)
+              .map(insertAndDelete)
 
-          idses.map(_.toInsertDelete) ~> transactInsertsDeletes ~> acsSink
-          lastOff ~> offsetSink
+            idses.map(_.toInsertDelete) ~> transactInsertsDeletes ~> acsSink
+            lastOff ~> offsetSink
 
-          ClosedShape
-        }
-      )
+            ClosedShape
+          }
+        )
 
-      val (acsQueue, lastOffsetFuture) = graph.run()
+        val (acsQueue, lastOffsetFuture) = graph.run()
+
+        for {
+          _ <- sinkCioSequence_(acsQueue)
+          offset0 <- connectionIOFuture(lastOffsetFuture)
+          offsetOrError <- offset0 max AbsoluteBookmark(absEnd.toDomain) match {
+            case ab @ AbsoluteBookmark(newOffset) =>
+              ContractDao
+                .updateOffset(parties, templateId, newOffset, offsets)
+                .map(_ => ab)
+            case LedgerBegin =>
+              fconn.pure(LedgerBegin)
+          }
+        } yield offsetOrError
+      }
+  }
 
-      for {
-        _ <- sinkCioSequence_(acsQueue)
-        offset0 <- connectionIOFuture(lastOffsetFuture)
-        offsetOrError <- offset0 max AbsoluteBookmark(absEnd.toDomain) match {
-          case ab @ AbsoluteBookmark(newOffset) =>
-            ContractDao
-              .updateOffset(parties, templateId, newOffset, offsets)
-              .map(_ => ab)
-          case LedgerBegin =>
-            fconn.pure(LedgerBegin)
+  private def debugLogAction[T, C](
+      actionDescription: String
+  )(block: => T)(implicit lc: LoggingContextOf[C]): T = {
+    if (logger.debug.isEnabled) {
+      val startTime = System.nanoTime()
+      logger.debug(s"Starting $actionDescription")
+      val result =
+        try {
+          block
+        } catch {
+          case e: Exception =>
+            lazy val endTime = System.nanoTime()
```

Comment:
I don't think there's any need to use `lazy` here. It'll only be evaluated once anyway, and only if needed, and it is not the kind of thing you'd want to memoise. Also it adds unnecessary synchronisation overhead to the runtime.
```suggestion
            val endTime = System.nanoTime()
```

Reply:
the synchronisation is what I want to ensure.... have better idea....will push it in a sec

Reply:
Done!!!